test/test_unary_ufuncs.py

```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
 
         self.assertEqual(actual, expected)
 
+    def _compare_out_variant_based_on_device(self, op, input, out):
+        out_dtype = out.dtype
+        if self.device_type == 'cuda':
+            # As input are dynamically casted
+            # in kernel, cast input to out_dtype
+            expected = op(input.to(out_dtype))
+            op(input, out=out)
+            self.assertEqual(out, expected)
+        else:
+            expected = op(input)
+            op(input, out=out)
+            # As output are casted to passed
+            # result dtype on CPU.
+            self.assertEqual(out, expected.to(out_dtype))
+
+    def _test_out_arg(self, op, input, output):
+        dtype = input.dtype
+        out_dtype = output.dtype
+        if dtype is out_dtype:
+            expected = op(input)
+            op(input, out=output)
+            self.assertEqual(output, expected)
+        else:
+            with self.assertRaises(RuntimeError):
+                op(input, out=output)
+
+    def _test_out_promote_int_to_float_op(self, op, input, output):
+        dtype = input.dtype
+        out_dtype = output.dtype
+        if out_dtype.is_floating_point and not dtype.is_complex:
+            self._compare_out_variant_based_on_device(op, input, output)
```

Comment:
Is there a test that sin(int, out=fp64) works? That is, the float32 output should be able to be cast to (weakly) wider types, like double and cfloat.